File: budget-pipeline/parsers/citi.py
"""
Citi CSV parser.

Headers: Status, Date, Description, Debit, Credit, Member Name

- Debit column populated (non-empty, non-zero) = charge → include
- Credit column populated = credit → parse_refunds() only
- Member Name column is NEVER stored (PII).
"""

import csv
from dateutil.parser import parse as parse_date

from categorizer import clean_merchant
from parsers.transaction import Transaction


def _read_rows(file_path: str) -> list[dict]:
    rows = []
    with open(file_path, encoding="utf-8", errors="replace") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)
    return rows


def _parse_amount(value: str) -> float:
    """Parse a dollar amount string, returning 0.0 if empty or invalid."""
    val = value.strip().replace(",", "").replace("$", "")
    if not val:
        return 0.0
    try:
        return float(val)
    except ValueError:
        return 0.0


def parse(file_path: str, account: str) -> list[Transaction]:
    """
    Parse Citi CSV. Returns debit (charge) transactions only.
    Drops Member Name column (PII).
    """
    rows = _read_rows(file_path)
    transactions: list[Transaction] = []

    for row in rows:
        raw_desc = row.get("Description", "").strip()
        if not raw_desc:
            continue

        try:
            txn_date = parse_date(row.get("Date", "").strip()).date()
        except Exception:
            continue

        debit_amount = _parse_amount(row.get("Debit", ""))

        if debit_amount > 0:
            merchant = clean_merchant(raw_desc)
            txn = Transaction(
                date=txn_date,
                description=raw_desc,
                merchant=merchant,
                amount=debit_amount,
                account=account,
                bank="citi",
                category="Misc",
                txn_type="expense",
            )
            transactions.append(txn)

    return transactions


def parse_refunds(file_path: str, account: str) -> list[Transaction]:
    """
    Parse Citi CSV. Returns credit transactions only.
    Drops Member Name column (PII).
    """
    rows = _read_rows(file_path)
    refunds: list[Transaction] = []

    for row in rows:
        raw_desc = row.get("Description", "").strip()
        if not raw_desc:
            continue

        try:
            txn_date = parse_date(row.get("Date", "").strip()).date()
        except Exception:
            continue

        credit_amount = _parse_amount(row.get("Credit", ""))

        if credit_amount > 0:
            merchant = clean_merchant(raw_desc)
            txn = Transaction(
                date=txn_date,
                description=raw_desc,
                merchant=merchant,
                amount=credit_amount,
                account=account,
                bank="citi",
                category="Misc",
                txn_type="expense",
            )
            refunds.append(txn)

    return refunds

File: web/lib/import/parsers/citi.ts
// TS port of budget-pipeline/parsers/citi.py.
// Headers: Status, Date, Description, Debit, Credit, Member Name
// Debit populated = charge; Credit populated = refund. Member Name dropped (PII).

import type { ParsedTransaction } from '../types'
import { readCsv, parseAmount } from '../csv'
import { parseDateISO } from '../date'
import { makeTxn } from './shared'

function collect(text: string, account: string, col: 'Debit' | 'Credit'): ParsedTransaction[] {
  const { rows } = readCsv(text)
  const out: ParsedTransaction[] = []

  for (const row of rows) {
    const rawDesc = (row['Description'] ?? '').trim()
    if (!rawDesc) continue

    const date = parseDateISO(row['Date'])
    if (!date) continue

    const amount = parseAmount(row[col])
    if (amount > 0) {
      out.push(makeTxn({ date, description: rawDesc, amount, account, bank: 'citi' }))
    }
  }

  return out
}

export function parse(text: string, account: string): ParsedTransaction[] {
  return collect(text, account, 'Debit')
}

export function parseRefunds(text: string, account: string): ParsedTransaction[] {
  return collect(text, account, 'Credit')
}
